### ISSUE TITLE
Odd behavior when calling C library function on windows

### ISSUE THREAD

**Issue by christopher-hesse** [ziglang/zig#8668]
I'm using zig 0.7.1 on windows, attempting to call a function from a `.lib` file.  The following zig program works:

```
const std = @import("std");

const c = @cImport({
    @cInclude("mkl.h");
});


pub fn main() anyerror!void {
    var a_matrix: f32 = 2.0;
    var b_matrix: f32 = 2.0;
    var c_matrix: f32 = 2.0;
    var lda : c.MKL_INT = 4;
    var ldb : c.MKL_INT = 4;
    var ldc : c.MKL_INT = 4;
    var m : c.MKL_INT = 1;
    var n : c.MKL_INT = 1;
    var k : c.MKL_INT = 1;
    // this program fails if these lines are uncommented
    // var alpha : f32 = 1.0;
    // var beta : f32 = 1.0;
    c.cblas_sgemm(
        c.CBLAS_LAYOUT.CblasRowMajor,
        c.CBLAS_TRANSPOSE.CblasNoTrans,
        c.CBLAS_TRANSPOSE.CblasNoTrans,
        m,
        n,
        k,
        1.0,
        &a_matrix,
        lda,
        &b_matrix,
        ldb,
        1.0,
        &c_matrix,
        ldc,
    );
    std.log.info("Output {}", .{c_matrix});
}
```

```
info: Output 6.0e+00
```

Uncommenting the two lines that declare variables causes this to fail:

```
Intel MKL ERROR: Parameter 9 was incorrect on entry to cblas_sgemm.
info: Output 2.0e+00
```

The following C program seems to work fine:

```
// https://software.intel.com/sites/products/documentation/doclib/mkl_sa/11/mkl_lapack_examples/c_bindings.htm
// cl test.c /Fatest.list /link mkl_intel_lp64_dll.lib mkl_core_dll.lib mkl_intel_thread_dll.lib libiomp5md.lib

#include <stdio.h>
#include <mkl.h>

int main() {
    float a_matrix = 2.0;
    float b_matrix = 2.0;
    float c_matrix = 2.0;
    MKL_INT lda = 4;
    MKL_INT ldb = 4;
    MKL_INT ldc = 4;
    MKL_INT m = 1;
    MKL_INT n = 1;
    MKL_INT k = 1;
    float alpha = 1.0;
    float beta = 1.0;
    CBLAS_LAYOUT layout = CblasRowMajor;
    CBLAS_TRANSPOSE transa = CblasNoTrans;
    CBLAS_TRANSPOSE transb = CblasNoTrans;
    cblas_sgemm(
        layout,
        transa,
        transb,
        m,
        n,
        k,
        alpha,
        &a_matrix,
        lda,
        &b_matrix,
        ldb,
        beta,
        &c_matrix,
        ldc
    );
    printf("output %f\n", c_matrix);
}
```

Looking at the generated assembly:

C
```
	mov	eax, DWORD PTR ldc$[rsp]
	mov	DWORD PTR [rsp+104], eax
	lea	rax, QWORD PTR c_matrix$[rsp]
	mov	QWORD PTR [rsp+96], rax
	movss	xmm0, DWORD PTR beta$[rsp]
	movss	DWORD PTR [rsp+88], xmm0
	mov	eax, DWORD PTR ldb$[rsp]
	mov	DWORD PTR [rsp+80], eax
	lea	rax, QWORD PTR b_matrix$[rsp]
	mov	QWORD PTR [rsp+72], rax
	mov	eax, DWORD PTR lda$[rsp]
	mov	DWORD PTR [rsp+64], eax
	lea	rax, QWORD PTR a_matrix$[rsp]
	mov	QWORD PTR [rsp+56], rax
	movss	xmm0, DWORD PTR alpha$[rsp]
	movss	DWORD PTR [rsp+48], xmm0
	mov	eax, DWORD PTR k$[rsp]
	mov	DWORD PTR [rsp+40], eax
	mov	eax, DWORD PTR n$[rsp]
	mov	DWORD PTR [rsp+32], eax
	mov	r9d, DWORD PTR m$[rsp]
	mov	r8d, DWORD PTR transb$[rsp]
	mov	edx, DWORD PTR transa$[rsp]
	mov	ecx, DWORD PTR layout$[rsp]
	call	cblas_sgemm
```

Zig working:

```
	.cv_loc	5 4 8 5
	vmovss	dword ptr [rbp + 44], xmm1
	.cv_loc	5 4 9 5
	vmovss	dword ptr [rbp + 40], xmm1
	.cv_loc	5 4 10 5
	vmovss	dword ptr [rbp + 36], xmm1
	.cv_loc	5 4 11 5
	mov	dword ptr [rbp + 24], 4
	.cv_loc	5 4 12 5
	mov	dword ptr [rbp + 20], 4
	.cv_loc	5 4 13 5
	mov	dword ptr [rbp + 16], 4
	.cv_loc	5 4 14 5
	mov	dword ptr [rbp + 12], 1
	.cv_loc	5 4 15 5
	mov	dword ptr [rbp + 8], 1
	.cv_loc	5 4 16 5
	mov	dword ptr [rbp + 4], 1
	.cv_loc	5 4 24 9
	mov	r9d, dword ptr [rbp + 12]
	.cv_loc	5 4 25 9
	mov	ecx, dword ptr [rbp + 8]
	.cv_loc	5 4 26 9
	mov	edx, dword ptr [rbp + 4]
	.cv_loc	5 4 29 9
	mov	r8d, dword ptr [rbp + 24]
	.cv_loc	5 4 31 9
	mov	r10d, dword ptr [rbp + 20]
	.cv_loc	5 4 34 9
	mov	r11d, dword ptr [rbp + 16]
	.cv_loc	5 4 20 18
	mov	esi, 101
	mov	dword ptr [rbp], ecx
	mov	ecx, esi
	mov	esi, 111
	mov	dword ptr [rbp - 4], edx
	mov	edx, esi
	mov	dword ptr [rbp - 8], r8d
	mov	r8d, esi
	mov	esi, dword ptr [rbp]
	mov	dword ptr [rsp + 32], esi
	mov	esi, dword ptr [rbp - 4]
	mov	dword ptr [rsp + 40], esi
	vmovss	dword ptr [rsp + 48], xmm0
	lea	rax, [rbp + 44]
.Ltmp79:
	mov	qword ptr [rsp + 56], rax
	mov	esi, dword ptr [rbp - 8]
	mov	dword ptr [rsp + 64], esi
	lea	rax, [rbp + 40]
.Ltmp80:
	mov	qword ptr [rsp + 72], rax
	mov	dword ptr [rsp + 80], r10d
	vmovss	dword ptr [rsp + 88], xmm0
	lea	rax, [rbp + 36]
.Ltmp81:
	mov	qword ptr [rsp + 96], rax
	mov	dword ptr [rsp + 104], r11d
	call	cblas_sgemm
```

Zig broken:

```
	.cv_loc	5 4 8 5
	vmovss	dword ptr [rbp + 60], xmm2
	.cv_loc	5 4 9 5
	vmovss	dword ptr [rbp + 56], xmm2
	.cv_loc	5 4 10 5
	vmovss	dword ptr [rbp + 52], xmm2
	.cv_loc	5 4 11 5
	mov	dword ptr [rbp + 40], 4
	.cv_loc	5 4 12 5
	mov	dword ptr [rbp + 36], 4
	.cv_loc	5 4 13 5
	mov	dword ptr [rbp + 32], 4
	.cv_loc	5 4 14 5
	mov	dword ptr [rbp + 28], 1
	.cv_loc	5 4 15 5
	mov	dword ptr [rbp + 24], 1
	.cv_loc	5 4 16 5
	mov	dword ptr [rbp + 20], 1
	.cv_loc	5 4 18 5
	vmovsd	qword ptr [rbp + 8], xmm1
	.cv_loc	5 4 19 5
	vmovsd	qword ptr [rbp], xmm1
	.cv_loc	5 4 24 9
	mov	r9d, dword ptr [rbp + 28]
	.cv_loc	5 4 25 9
	mov	ecx, dword ptr [rbp + 24]
	.cv_loc	5 4 26 9
	mov	edx, dword ptr [rbp + 20]
	.cv_loc	5 4 29 9
	mov	r8d, dword ptr [rbp + 40]
	.cv_loc	5 4 31 9
	mov	r10d, dword ptr [rbp + 36]
	.cv_loc	5 4 34 9
	mov	r11d, dword ptr [rbp + 32]
	.cv_loc	5 4 20 18
	mov	esi, 101
	mov	dword ptr [rbp - 4], ecx
	mov	ecx, esi
	mov	esi, 111
	mov	dword ptr [rbp - 8], edx
	mov	edx, esi
	mov	dword ptr [rbp - 12], r8d
	mov	r8d, esi
	mov	esi, dword ptr [rbp - 4]
	mov	dword ptr [rsp + 32], esi
	mov	esi, dword ptr [rbp - 8]
	mov	dword ptr [rsp + 40], esi
	vmovss	dword ptr [rsp + 48], xmm0
	lea	rax, [rbp + 60]
.Ltmp79:
	mov	qword ptr [rsp + 56], rax
	mov	esi, dword ptr [rbp - 12]
	mov	dword ptr [rsp + 64], esi
	lea	rax, [rbp + 56]
.Ltmp80:
	mov	qword ptr [rsp + 72], rax
	mov	dword ptr [rsp + 80], r10d
	vmovss	dword ptr [rsp + 88], xmm0
	lea	rax, [rbp + 52]
.Ltmp81:
	mov	qword ptr [rsp + 96], rax
	mov	dword ptr [rsp + 104], r11d
	call	cblas_sgemm
```

If I'm reading this correctly, r9d is being set to different values in the two zig cases.  Am I missing some option that is required to make this work, or am I calling the function incorrectly somehow?

**Comment by christopher-hesse** [comment]
The output of `zig build --verbose-llvm-ir` looks reasonable to me, though I don't know LLVM IR:

```
  %2 = load i32, i32* %m, align 4, !dbg !4706
  %3 = load i32, i32* %n, align 4, !dbg !4707
  %4 = load i32, i32* %k, align 4, !dbg !4708
  %5 = load i32, i32* %lda, align 4, !dbg !4709
  %6 = load i32, i32* %ldb, align 4, !dbg !4710
  %7 = load i32, i32* %ldc, align 4, !dbg !4711
  call void @cblas_sgemm(i32 101, i32 111, i32 111, i32 %2, i32 %3, i32 %4, float 1.000000e+00, float* %a_matrix, i32 %5, float* %b_matrix, i32 %6, float 1.000000e+00, float* %c_matrix, i32 %7), !dbg !4712
```

**Comment by christopher-hesse** [comment]
Here's the generated definition from `zig build --verbose-cimport` 

```
pub extern fn cblas_sgemm(Layout: CBLAS_LAYOUT, TransA: CBLAS_TRANSPOSE, TransB: CBLAS_TRANSPOSE, M: c_int, N: c_int, K: c_int, alpha: f32, A: [*c]const f32, lda: c_int, B: [*c]const f32, ldb: c_int, beta: f32, C: [*c]f32, ldc: c_int) void;
```

**Comment by christopher-hesse** [comment]
Because the variable declarations alter this, it could be the alignment of arguments on the stack.

**Comment by alexnask** [comment]
@christopher-hesse 
I cannot reproduce this with a zig master build, can you confirm on your end as well?

**Comment by christopher-hesse** [comment]
Thanks for looking at this @alexnask, I should have tried master earlier, but in my case it doesn't seem to help.  Calling my own C function with similar args works fine, but calling `cblas_sgemm` does not.   It's very possible I am mis-specifying something here, but it did seem to work without issue when I used the C program above.

With the master version of zig `zig-windows-x86_64-0.8.0-dev.2082+e9e91b4ed` I got the following issues:

* `lld-link: error: could not open 'tbb12.lib': no such file or directory` => resolved by adding ```const tbb_path = "C:\\Program Files (x86)\\Intel\\oneAPI\\tbb\\2021.2.0\\";
    exe.addLibPath(tbb_path ++ "lib\\intel64\\vc14");``` to build.zig
* `The following command exited with error code 3221225781 (expected 0):` => resolved by copying `C:\Program Files (x86)\Intel\oneAPI\tbb\2021.2.0\redist\intel64\vc14\tbb12.dll` to `zig-out\bin`
* Run example program with no unused variables:
```
Intel MKL ERROR: Parameter 9 was incorrect on entry to cblas_sgemm.
info: Output 2.0e+00 c_int@94881ff7b0
```

* Run example program with unused u64 variable:
```
Segmentation fault at address 0x277da00008
???:?:?: 0x7ff730a1d8fd in ??? (mkl_core.lib)
The following command exited with error code 2147483651 (expected 0):
```

Were you able to call `cblas_sgemm` specifically?  Something may be wrong with my configuration if so.
